Fix type juggling notes for list() and array destructuring

tgr opened the pull request:
Update notes about the handling of missing keys and non-arrays during array destructuring to match the behavior seen on 3v4l.org, which doesn't issue a warning or notice in some cases where the documentation claims it does. Also update list() docs to provide the same information.

list():
* missing key: https://3v4l.org/bYBt6
* destructuring an integer: https://3v4l.org/7lNhK
* a boolean: https://3v4l.org/p5LDd
* a string: https://3v4l.org/KHSAH
* null: https://3v4l.org/lk1iu

array notation:
* missing key: https://3v4l.org/KCiTC
* destructuring an integer: https://3v4l.org/louVG
* a boolean: https://3v4l.org/SWL5L
* a string: https://3v4l.org/aK90L
* null: https://3v4l.org/Y8cOY

Fixes #3640.

MatmaRex commented on language/types/array.xml:
```diff
@@ -391,10 +391,7 @@ $secondElement = getArray()[1];
-     Array dereferencing a scalar value which is not a <type>string</type>
-     yields &null;. Prior to PHP 7.4.0, that did not issue an error message.
-     As of PHP 7.4.0, this issues <constant>E_NOTICE</constant>;
-     as of PHP 8.0.0, this issues <constant>E_WARNING</constant>.
```
This was about array dereferencing (`$a = 42; $a[0]`) not destructuring (`[$a] = 42;`), and I think it was correct. Only destructuring weirdly yields null with no warning.

MatmaRex commented on reference/array/functions/list.xml:
```diff
@@ -25,6 +25,20 @@
+  </note>
+  <note>
+   <para>
+    Attempting to access element of a scalar value yields &null;.
```
I would say: "Attempting to unpack a scalar value assigns null to all variables." to be consistent with the existing wording.

There's also already a note above: "Strings can not be unpacked" which should probably say "Only arrays and objects that implement ArrayAccess can be unpacked".

And an example below, where the comment could be clarified:
```
// list() doesn't work with strings
list($bar) = "abcde";
var_dump($bar); // NULL
```

tgr replied:
Updated the wording and the other note. IMO documenting behavior when misused is not worth the space for adding an example.

DanielEScherzer reviewed (changes requested):
Sorry it has taken us so long to review this; thanks for the patch, one small issue

MatmaRex commented:
This probably needs an update now for changes in PHP 8.5: https://wiki.php.net/rfc/warnings-php-8-5#destructuring_non-array_values

tgr commented:
Sorry, I forgot about this PR completely. Thanks for the feedback, rebased and updated.

> This probably needs an update now for changes in PHP 8.5: https://wiki.php.net/rfc/warnings-php-8-5#destructuring_non-array_values

I think the docs generally don't include details about not-yet-released PHP versions?

DanielEScherzer commented:
> Sorry, I forgot about this PR completely. Thanks for the feedback, rebased and updated.
> 
> > This probably needs an update now for changes in PHP 8.5: https://wiki.php.net/rfc/warnings-php-8-5#destructuring_non-array_values
> 
> I think the docs generally don't include details about not-yet-released PHP versions?

Yeah, we don't need to start worrying about that yet
